Source/Core/InputCommon/InputProfile.cpp

```diff
@@ -0,0 +1,187 @@
+// Copyright 2018 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "Common/FileSearch.h"
+#include "Common/FileUtil.h"
+
+#include "Core/Config/WiimoteInputSettings.h"
+#include "Core/Core.h"
+#include "Core/HW/Wiimote.h"
+#include "Core/HotkeyManager.h"
+
+#include "InputCommon/ControllerInterface/ControllerInterface.h"
+#include "InputCommon/InputConfig.h"
+#include "InputCommon/InputProfile.h"
+
+#include <algorithm>
+#include <iterator>
+
+namespace
+{
+  enum class CycleDirection : int
+  {
+    Forward = 1,
+    Backward = -1
+  };
+
+  int s_wiimote_profile_index = 0;
```

Comment:
Please don't make this file-static-based. All of this behavior should probably be put in a class, rather than a flat namespace, and be driven by the front-end in some form.

Reply:
I did this mostly because I'm trying to A) keep everything encompassed in one place.  B) Was trying to avoid exposing too much to the UI.

The original implementation of this logic had the index in the hotkey controller interface which I didn't agree with.

Where do things typically get stored?  Even if this was a class, I'm not sure how I feel about the UI 'owning' it directly.

Reply:
If hotkeys are driven by the UI, I really don't see how controller profiles are an exception to this, given there needs to be interaction from the UI to manipulate profiles, therefore the UI owning it would make the most sense here. Realistically may do what you want in terms of where it actually goes as long as it's not static-based like this. This adds "yet another thing to undo/fix" in terms of working towards libdolphin otherwise.

Reply:
Yeah hotkeys are driven by the UI but I don't really think they should be.

Ideally, I feel like everything should be separated out.  Hotkey ownership would be managed by each type (ex: graphics hotkeys owned by a graphics class, inputs hotkeys owned by a input class, etc).  The UI would have specific touch points into these classes to get at the data it needs or to call the actions it needs to run.  Each class would have its own responsibility and ownership.  UI would never know about specific hotkey interactions, it would just say `handle_hotkeys()`  If you know about the MVVM paradigm, I feel like we're breaking it at the moment.

All that being said, if putting it in a class that is then owned by the UI is the quickest way to get this solved, then I'll likely go for that.

And `libdolphin`?  That's the first I've heard of it.  What would that be used for?

Reply:
It's the quickest way to get it approved (by me, if that matters).

also libdolphin would allow the core to be use separately from the project (i.e. for some other application purpose; think something along the lines of [Unicorn](https://www.unicorn-engine.org/)

Reply:
Getting the code approved by you is always a goal for me because I know you strive for high-quality code and that's something I aim for as well.  Thank you for taking the time to review.  I will take a look at this in the next day or two.

Also, appreciate the insight into `libdolphin`, very interesting!